provider/core/src/data_provider.rs

```diff
@@ -146,6 +151,41 @@ where
 {
 }
 
+/// A wrapper type that wraps either an [`Rc`](alloc::rc::Rc) or an
+/// [`Arc`](alloc::sync::Arc), depending on the "sync" feature. Create
+/// this from a `&[u8]`.
+#[derive(Clone)]
+pub struct RcWrap(
+    #[cfg(not(feature = "sync"))]
+    alloc::rc::Rc<[u8]>,
+    #[cfg(feature = "sync")]
+    alloc::sync::Arc<[u8]>,
+);
+
+impl core::ops::Deref for RcWrap {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+unsafe impl CloneableCart for RcWrap {}
+unsafe impl stable_deref_trait::StableDeref for RcWrap {}
```

Comment:
Nit: Always include safety docs. Each trait should have a section that tells you what invariants need to be upheld.